src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -1120,26 +1102,30 @@ mongoc_topology_select_server_id (mongoc_topology_t *topology,
          error->domain = MONGOC_ERROR_SERVER_SELECTION;
          error->code = MONGOC_ERROR_SERVER_SELECTION_FAILURE;
       }
-      bson_mutex_unlock (&topology->mutex);
-      return 0;
+      server_id = 0;
+      goto done;
    }
 
-   if (topology->description.type == MONGOC_TOPOLOGY_LOAD_BALANCED) {
-      bson_mutex_unlock (&topology->mutex);
-      return _mongoc_topology_select_server_id_loadbalanced (topology, error);
+   if (td.ptr->type == MONGOC_TOPOLOGY_LOAD_BALANCED) {
+      server_id =
+         _mongoc_topology_select_server_id_loadbalanced (topology, error);
+      goto done;
    }
 
-   bson_mutex_unlock (&topology->mutex);
-
-   heartbeat_msec = topology->description.heartbeat_msec;
+   heartbeat_msec = td.ptr->heartbeat_msec;
    local_threshold_ms = topology->local_threshold_msec;
    try_once = topology->server_selection_try_once;
    loop_start = loop_end = bson_get_monotonic_time ();
    expire_at =
       loop_start + ((int64_t) topology->server_selection_timeout_msec * 1000);
 
    if (topology->single_threaded) {
-      _mongoc_topology_description_monitor_opening (&topology->description);
+      if (!td.ptr->opened) {
+         mc_tpld_modification tdmod = mc_tpld_modify_begin (topology);
+         _mongoc_topology_description_monitor_opening (tdmod.new_td);
+         mc_tpld_modify_commit (tdmod);
+         mc_tpld_renew_ref (&td, topology);
```

Comment:
I think this is a good question to debate on whether it is preferable to have uniformity, or to just use the direct method in the single-threaded case.

This case is fairly benign, as `monitor_opening` doesn't add/remove/tweak any server descriptions, so it is safe to do the "unsafe" thing. In other cases it may be that a caller has a live pointer to a server description which might be disappeared out from under them.